src/python/pants/backend/go/goals/generate.py

```diff
@@ -0,0 +1,285 @@
+        if args[0] == "-command":
+            if len(args) <= 1:
+                raise ValueError(
+                    f"{go_file}:{line_num}: -command syntax used but no command name specified"
+                )
+            cmd_shorthand[args[1]] = tuple(args[2:])
+            continue
+
+        # Replace any shorthand command with the previously-stored arguments.
+        if args[0] in cmd_shorthand:
+            args = [*cmd_shorthand[args[0]], *args[1:]]
+
+        # TODO: Substitute environment variables.
+
+        # If the program calls for `go`, then use the full path to the `go` binary in the GOROOT.
+        if args[0] == "go":
+            args[0] = os.path.join(goroot.path, "bin", "go")
+
+        generate_env = {
+            "GOOS": goroot.goos,
+            "GOARCH": goroot.goarch,
+            "GOFILE": go_file,
+            "GOLINE": str(line_num),
+            "GOPACKAGE": analysis.name,
+            "GOROOT": goroot.path,
+            "DOLLAR": "$",
+        }
+
+        for i, arg in enumerate(args):
+            args[i] = _expand_env(arg, generate_env)
+
+        # Invoke the subprocess and store its output for use as input root of next command (if any).
+        result = await Get(
+            ProcessResult,
+            Process(
+                argv=args,
+                input_digest=digest,
+                working_directory=dir_path,
+                output_directories=["."],
+                env={
+                    **generate_env,
+                    **env,
+                },
+                description=f"Process `go generate` directives in file: {os.path.join(dir_path, go_file)}",
+            ),
+        )
+        digest = await Get(Digest, AddPrefix(result.output_digest, dir_path))
+
+    return digest
+
+
+@rule
+async def run_go_package_generators(
+    request: RunPackageGeneratorsRequest, goroot: GoRoot, subsystem: GoGenerateGoalSubsystem
+) -> RunPackageGeneratorsResult:
+    fallible_analysis, env = await MultiGet(
+        Get(
+            FallibleFirstPartyPkgAnalysis,
+            FirstPartyPkgAnalysisRequest(request.address, extra_build_tags=("generate",)),
+        ),
+        Get(EnvironmentVars, EnvironmentVarsRequest(subsystem.env_vars)),
+    )
+    if not fallible_analysis.analysis:
+        raise ValueError(f"Analysis failure for {request.address}: {fallible_analysis.stderr}")
+    analysis = fallible_analysis.analysis
+    dir_path = analysis.dir_path if analysis.dir_path else "."
+
+    fallible_pkg_digest = await Get(
+        FallibleFirstPartyPkgDigest, FirstPartyPkgDigestRequest(request.address)
+    )
+    if fallible_pkg_digest.pkg_digest is None:
+        raise ValueError(
+            f"Unable to obtain digest for {request.address}: {fallible_pkg_digest.stderr}"
+        )
+    pkg_digest = fallible_pkg_digest.pkg_digest
+
+    # Scan each Go file in the package for generate directives. Process them sequentially so that an error in
```

Comment:
How is the ordering of `analysis.go_files` determined? Does it matter which is used first?

Reply:
I believe analysis sorts them in alphabetical order just like Go toolchain does when it analyzes. For `go generate` purposes, the order does matter but it should just follow whatever analysis returns.